### ISSUE TITLE
[Windows] TextField.TextChanged fires previous value 

### ISSUE THREAD

**Issue by slimCODE** [migueldeicaza/gui.cs#1573]
Given the following code:
```csharp
using Terminal.Gui;

Application.Init();

var input = new TextField()
{
    X = 1,
    Y = 1,
    Width = Dim.Fill(2),
    Height = 1
};

var output = new Label()
{
    X = 1,
    Y = 3,
    Width = Dim.Fill(2),
    Height = 1
};

input.TextChanged += text =>
{
    System.Diagnostics.Debug.WriteLine(text);
    output.Text = text;
};

var window = new Window()
{
    X = 0,
    Y = 0,
    Width = Dim.Fill(),
    Height = Dim.Fill()
};

window.Add(input, output);

Application.Run(window);
```

The debug output will always display the content of the TextField prior to the last key pressed.

![image](https://user-images.githubusercontent.com/13796362/149053058-0388bb51-3b07-4ea7-836c-0d47b169c359.png)

**Comment by tznind** [comment]
This appears to be by design:

https://github.com/migueldeicaza/gui.cs/blob/646a88e4bac25885425bc0382218fd25c56d3157/Terminal.Gui/Views/TextField.cs#L48-L50

You will need to reference `input.Text` instead of the value passed into the event (which is the 'old value')

**Comment by BDisp** [comment]
> You will need to reference `input.Text` instead of the value passed into the event (which is the 'old value').

That is right, it must reference the `input.Text`.


> Having the old value is handy if you want to do validation for example and reset the old value if the new value is invalid.

No, it's only to know what is the old value after the changes are done. For validation and cancel the new value there is the cancelable `TextChanging` event, which is fired before thee `TextChanged` event.


> Ideally there would be a `TextChangedEvent` class or something which contained both the old and new values and the `sender` but we try not to make breaking changes without a good reason.

There was a discussion a long time ago to decide which event type we should use in the `Terminal.Gui`. Before, it only used `Action` object without the `event` declaration, which allowed declared the events inside the views constructors. Due the limitations of this type, we have discussed if we should use the `event Action` or the `event EventHandler`. Both will forced to change the events declarations in all projects, but we needed to take a decision. Personally, I preferred the `event EventHandler`, but due the fact that the `sender` argument not always needs to being used, the majority decided for the `event Action`. So, if sender is needed to be used, the better solution is implementing the `event EventHandler` for future views implementations and improvements. Passing the `sender` in the `EventArgs` implies creating a `EventArgs` class, even for those events do not have any arguments or only have one argument. When one day is needed to change this and breaking changes, the better solution is using the `event EventHandler` for all.
